# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/mac/GeckoTextMarker.mm
# Commit: 5972532c7737
# Full Hash: 5972532c773757211b0d30ed34a3849bae0dc5a9
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2020-07-22 04:32:41
# Regressor Bug: 1653421
# File Overlap Count: 2
# Description:
#   Bug 1653421 - Part 3: Don't normalize marker beyond editable root. r=morgan
#   
#   When in an editable container, text markers are expected to not go past or before the editable root.
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D84055
# ==============================================================================

diff -r 1d37ebf3d20c -r 5972532c7737 accessible/mac/GeckoTextMarker.mm
--- a/accessible/mac/GeckoTextMarker.mm	Tue Jul 21 23:02:55 2020 +0000
+++ b/accessible/mac/GeckoTextMarker.mm	Tue Jul 21 23:02:57 2020 +0000
@@ -100,6 +100,24 @@
   return false;
 }
 
+bool GeckoTextMarker::IsEditableRoot() {
+  uint64_t state =
+      mContainer.IsProxy() ? mContainer.AsProxy()->State() : mContainer.AsAccessible()->State();
+  if ((state & states::EDITABLE) == 0) {
+    return false;
+  }
+
+  AccessibleOrProxy parent = mContainer.Parent();
+  if (parent.IsNull()) {
+    // Not sure when this can happen, but it would technically be an editable root.
+    return true;
+  }
+
+  state = parent.IsProxy() ? parent.AsProxy()->State() : parent.AsAccessible()->State();
+
+  return (state & states::EDITABLE) == 0;
+}
+
 void GeckoTextMarker::NormalizeNext() {
   if (AtEnd()) {
     // If this is the end of the current container, mutate to its parent's
@@ -109,12 +127,14 @@
                                               : mContainer.AsAccessible()->EndOffset();
 
     if (endOffset != 0) {
-      mContainer = mContainer.Parent();
-      mOffset = endOffset;
+      if (!IsEditableRoot()) {
+        mContainer = mContainer.Parent();
+        mOffset = endOffset;
 
-      // Call NormalizeNext recursively to get top-most link if at the end of one,
-      // or innermost link if at the beginning.
-      NormalizeNext();
+        // Call NormalizeNext recursively to get top-most link if at the end of one,
+        // or innermost link if at the beginning.
+        NormalizeNext();
+      }
     }
   } else {
     AccessibleOrProxy link;
@@ -146,12 +166,14 @@
                                                 : mContainer.AsAccessible()->StartOffset();
 
     if (startOffset != 0) {
-      mContainer = mContainer.Parent();
-      mOffset = startOffset;
+      if (!IsEditableRoot()) {
+        mContainer = mContainer.Parent();
+        mOffset = startOffset;
 
-      // Call NormalizePrevious recursively to get top-most link if at the start of one,
-      // or innermost link if at the end.
-      NormalizePrevious();
+        // Call NormalizePrevious recursively to get top-most link if at the start of one,
+        // or innermost link if at the end.
+        NormalizePrevious();
+      }
     }
   } else {
     AccessibleOrProxy link;
